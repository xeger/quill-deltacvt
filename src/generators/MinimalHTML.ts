import BaseHTML, { EmbedFormatter, TextFormatter } from './BaseHTML';

const BODY_STYLE =
  'color:#303030;font-weight:400;white-space:pre-wrap;font-family:sans-serif';

/**
 * Inline CSS styles for selected line formats.
 */
const LINE_STYLES = {
  list: 'margin:0px;list-style-position:inside;padding-left:1.5rem',
};

/**
 * HTML generators for non-textual content (e.g. images)
 * in a chunk. Understands width, height and float.
 * @see https://github.com/xeger/quill-image for float support
 */
const EMBEDS: Record<string, EmbedFormatter> = {
  image(src, attributes) {
    const { float, width, height } = attributes;
    const optWidth = width ? ` width="${width}"` : '';
    const optHeight = height ? ` height="${height}"` : '';
    const optStyle = float ? ` style="float:${float}"` : '';

    return `<img src="${src}"${optWidth}${optHeight}${optStyle}/>`;
  },
};

/**
 * HTML generators for character formats. Line formats are context-sensitive
 * and need to be handled as part of the generation algorithm.
 */
const TEXTS: Record<string, TextFormatter> = {
  bold: (span) => {
    span.styles.push('font-weight:bold');
  },
  color: (span, color) => {
    span.styles.push(`color:${color}`);
  },
  font: (span, font) => {
    span.styles.push(`font-family:${font}`);
  },
  italic: (span) => {
    span.styles.push('font-style:italic');
  },
  link: (span, href) => {
    span.text = `<a href="${href}">${span.text}</a>`;
  },
  size: (span, size) => {
    span.styles.push(`font-size:${size}`);
  },
  underline: (span) => {
    span.styles.push('text-decoration:underline');
  },
};

export interface Options {
  strict?: true;
}

interface Line {
  align?: string;
  list?: string;
  text: string;
}

/**
 * Generator that outputs HTML approximating the visual style of Quill's
 * Parchment formats, but with notable differences:
 *   - the `<div>` tag is used instead of `<p>` for paragraphs
 *   - all styling is inline (no classes; no stylesheet)
 *
 * The resulting HTML is minimal in size, self contained with no need
 * for external stylesheets, minimally influenced by user-agent stylesheets,
 * and looks visually similar to Quill output in modern browsers, although
 * structurally quite different.
 *
 * To achieve this look, the generator applies a style to the `body` element
 * when you `wrap()`, which affects all of the inner content.
 *
 * If you generate fragments, make sure to apply a similar style (especially
 * a `white-space: pre-wrap`) to an enclosing tag, else the HTML will look
 * completely wrong!
 */
export default class MinimalHTML extends BaseHTML {
  embedFormatters: Record<string, EmbedFormatter> = { ...EMBEDS };
  strict?: true;
  textFormatters: Record<string, TextFormatter> = { ...TEXTS };

  constructor(options: Options = {}) {
    super(EMBEDS, TEXTS, options);
  }

  formatLine(current: Line, prior: Line): string {
    const { align, list, text } = current;
    const style = align ? ` style="text-align: ${align}"` : '';

    let html: string;
    if (list) html = `<li${style}>${text}</li>`;
    else if (text) html = `<div${style}>${text}</div>`;
    else html = '';

    if (list !== prior?.list) {
      const prefix: string[] = [];
      if (prior?.list === 'bullet') prefix.push(`</ul>`);
      else if (prior?.list === 'ordered') prefix.push(`</ol>`);
      if (list === 'bullet') prefix.push(`<ul style="${LINE_STYLES.list}">`);
      else if (list === 'ordered')
        prefix.push(`<ol style="${LINE_STYLES.list}">`);
      html = `${prefix.join('')}${html}`;
    }

    return html;
  }

  isLineFormat(format: string): boolean {
    switch (format) {
      case 'align':
      case 'list':
        return true;
      default:
        return false;
    }
  }

  wrap(fragment: string): string {
    return `<!DOCTYPE html><html><body style="${BODY_STYLE}">${fragment}</body></html>`;
  }
}
