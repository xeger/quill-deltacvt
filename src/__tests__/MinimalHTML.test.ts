import { generateFragment } from '..';
import MinimalHTML from '../generators/MinimalHTML';

const UNKNOWN_ATTRIBUTE = [{ attributes: { banana: true }, insert: 'hi' }];

const UNKNOWN_EMBED = [{ insert: { banana: 'banana' } }];

describe('generators/MinimalHTML', () => {
  it('handles unknown content', () => {
    const g = new MinimalHTML();
    expect(generateFragment(UNKNOWN_ATTRIBUTE, g)).toEqual('hi');
    expect(generateFragment(UNKNOWN_EMBED, g)).toEqual('');
  });

  describe('options', () => {
    describe('strict mode', () => {
      const g = new MinimalHTML({ strict: true });
      it('gives helpful embed messages', () => {
        expect(() => generateFragment(UNKNOWN_EMBED, g)).toThrowError(
          'Unknown embed: banana'
        );
      });

      it('gives helpful attribute messages', () => {
        expect(() => generateFragment(UNKNOWN_ATTRIBUTE, g)).toThrowError(
          'Unknown attribute: banana'
        );
      });
    });
  });
});
