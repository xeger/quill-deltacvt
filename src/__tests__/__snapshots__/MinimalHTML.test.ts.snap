// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generators/MinimalHTML algorithm escapes HTML syntax 1`] = `
"<!DOCTYPE html><html><body style=\\"color:#303030;font-weight:400;white-space:pre-wrap;font-family:sans-serif\\"><div>&lt;script&gt;alert(&quot;xss&quot;);&lt;/script&gt;
</div><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>&lt;script&gt;alert(&quot;xss&quot;);&lt;/script&gt;
</li><li style=\\"text-align: right\\">&lt;script&gt;alert(&quot;xss&quot;);&lt;/script&gt;
</li><li>not xss &amp; okay
</li><li style=\\"text-align: center\\">not xss &amp; okay
</li></ul><div>
</div></body></html>"
`;

exports[`generators/MinimalHTML algorithm handles fonts 1`] = `
"<!DOCTYPE html><html><body style=\\"color:#303030;font-weight:400;white-space:pre-wrap;font-family:sans-serif\\"><div><span style=\\"font-family:'courier'\\">Hello</span>, <span style=\\"font-family:'serif'\\">world</span>; do <span style=\\"font-weight:bold;font-family:'monospace'\\">you</span> like my fonts?
</div></body></html>"
`;

exports[`generators/MinimalHTML algorithm handles images 1`] = `
"<!DOCTYPE html><html><body style=\\"color:#303030;font-weight:400;white-space:pre-wrap;font-family:sans-serif\\"><div><img src=\\"https://quilljs.com/assets/images/logo.svg\\" width=\\"140\\" height=\\"38\\"/>
</div><div>Sized image.
</div><div>
</div><div><img src=\\"https://raw.githubusercontent.com/quilljs/quill/master/assets/favicon.png\\"/>
</div><div>Left-aligned image.
</div><div>
</div><div style=\\"text-align: center\\"><img src=\\"https://raw.githubusercontent.com/quilljs/quill/master/assets/favicon.png\\"/>
</div><div style=\\"text-align: center\\">Center-aligned image.
</div><div style=\\"text-align: center\\">
</div><div style=\\"text-align: right\\"><img src=\\"https://raw.githubusercontent.com/quilljs/quill/master/assets/favicon.png\\"/>
</div><div style=\\"text-align: right\\">Right-aligned image.
</div><div>
</div><div style=\\"text-align: center\\"><img src=\\"https://quilljs.com/assets/images/logo.svg\\" width=\\"140\\" height=\\"38\\"/>
</div><div style=\\"text-align: center\\">Sized and centered image.
</div></body></html>"
`;

exports[`generators/MinimalHTML algorithm handles line formats within the same DeltaOperation 1`] = `
"<!DOCTYPE html><html><body style=\\"color:#303030;font-weight:400;white-space:pre-wrap;font-family:sans-serif\\"><div style=\\"text-align: center\\">Centered.
</div><div style=\\"text-align: center\\">Centered <span style=\\"font-weight:bold;color:#b2b200\\">with</span> <span style=\\"text-decoration:underline\\"><a href=\\"http://quilljs.com\\">some</a></span> formatting.
</div><div style=\\"text-align: right\\">Right aligned.
</div><div style=\\"text-align: right\\"><span style=\\"color:#66b966\\">Right</span> <span style=\\"font-style:italic;text-decoration:underline\\">aligned</span> with <span style=\\"font-weight:bold\\">some</span> formatting.
</div><div style=\\"text-align: justify\\">
</div><div style=\\"text-align: justify\\"><span style=\\"font-weight:bold\\">Justified</span> so the text nicely fills the line. <span style=\\"color:#f06666\\">Justified</span> <span style=\\"color:#66a3e0\\">so</span> <span style=\\"color:#c285ff\\">the</span> text nicely fills the line. Justified so the text nicely fills the line.
</div><div style=\\"text-align: justify\\">Also justified.
</div><div>
</div><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Unordered
</li><li>List
</li></ul><div>
</div><ol style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Ordered
</li><li>List
</li></ol><div>
</div><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Mixed
</li></ul><ol style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>List
</li></ol><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Weirdly
</li></ul><ol style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Laid Out
</li><li>With Numbers
</li></ol><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Interspersed
</li></ul><div>
</div><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>list
</li><li style=\\"text-align: center\\">with
</li><li style=\\"text-align: right\\">alignment
</li><li style=\\"text-align: justify\\">and justification in case the list items get stupendously enormously long
</li></ul></body></html>"
`;

exports[`generators/MinimalHTML algorithm handles lists 1`] = `
"<!DOCTYPE html><html><body style=\\"color:#303030;font-weight:400;white-space:pre-wrap;font-family:sans-serif\\"><div>Normal text to start.
</div><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>chalk
</li><li>
</li><li>cheese
</li></ul><div>Now an ordered list:
</div><ol style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>apples
</li><li style=\\"text-align: right\\">oranges
</li><li>pears
</li><li>
</li></ol><div>And finally, adjacent lists:
</div><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>please
</li><li>excuse
</li><li>my
</li></ul><ol style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>dear
</li><li>aunt
</li><li>sally
</li></ol></body></html>"
`;

exports[`generators/MinimalHTML algorithm handles standard Quill formats 1`] = `
"<!DOCTYPE html><html><body style=\\"color:#303030;font-weight:400;white-space:pre-wrap;font-family:sans-serif\\"><div>Plain.  <span style=\\"font-weight:bold\\">Bold. </span><span style=\\"font-style:italic\\">Italic. </span><span style=\\"text-decoration:underline\\">Underline.</span> <span style=\\"color:#9933ff\\">Colored.</span> <a href=\\"https://quilljs.com\\">Link.</a>
</div><div><span style=\\"font-size:x-small\\">Small, </span>normal, <span style=\\"font-size:x-large\\">large, </span><span style=\\"font-size:xx-large\\">huge!</span>
</div><div>Normal family. <span style=\\"font-family:'monospace'\\">Monospace family.</span> <span style=\\"font-family:'serif'\\">Serif family.</span>
</div><div>Left aligned.
</div><div>Left <span style=\\"font-weight:bold\\">aligned</span> with <span style=\\"font-style:italic;text-decoration:underline\\">some</span> <span style=\\"color:#e60000\\">formatting</span>.
</div><div style=\\"text-align: center\\">Centered.
</div><div style=\\"text-align: center\\">Centered <span style=\\"font-weight:bold;color:#b2b200\\">with</span> <span style=\\"text-decoration:underline\\"><a href=\\"http://quilljs.com\\">some</a></span> formatting.
</div><div style=\\"text-align: right\\">Right aligned.
</div><div style=\\"text-align: right\\"><span style=\\"color:#66b966\\">Right</span> <span style=\\"font-style:italic;text-decoration:underline\\">aligned</span> with <span style=\\"font-weight:bold\\">some</span> formatting.
</div><div style=\\"text-align: justify\\">
</div><div style=\\"text-align: justify\\"><span style=\\"font-weight:bold\\">Justified</span> so the text nicely fills the line. <span style=\\"color:#f06666\\">Justified</span> <span style=\\"color:#66a3e0\\">so</span> <span style=\\"color:#c285ff\\">the</span> text nicely fills the line. Justified so the text nicely fills the line.
</div><div style=\\"text-align: justify\\">Also justified.
</div><div>
</div><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Unordered
</li><li>List
</li></ul><div>
</div><ol style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Ordered
</li><li>List
</li></ol><div>
</div><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Mixed
</li></ul><ol style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>List
</li></ol><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Weirdly
</li></ul><ol style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Laid Out
</li><li>With Numbers
</li></ol><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>Interspersed
</li></ul><div>
</div><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>list
</li><li style=\\"text-align: center\\">with
</li><li style=\\"text-align: right\\">alignment
</li><li style=\\"text-align: justify\\">and justification in case the list items get stupendously enormously long
</li></ul><div>
</div></body></html>"
`;

exports[`generators/MinimalHTML algorithm handles styled lists 1`] = `
"<!DOCTYPE html><html><body style=\\"color:#303030;font-weight:400;white-space:pre-wrap;font-family:sans-serif\\"><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li><span style=\\"font-weight:bold;text-decoration:underline\\">lots</span> <span style=\\"color:#9933ff\\">of</span> <span style=\\"font-weight:bold;font-style:italic;text-decoration:underline\\"><a href=\\"http://google.com\\">style</a></span> <span style=\\"font-style:italic\\">happening</span>
</li><li>lots <span style=\\"font-weight:bold\\">of</span> <span style=\\"color:#e60000\\">function</span> too
</li><li>gotta love it all 
</li></ul></body></html>"
`;

exports[`generators/MinimalHTML algorithm handles trivial aligned text 1`] = `
"<!DOCTYPE html><html><body style=\\"color:#303030;font-weight:400;white-space:pre-wrap;font-family:sans-serif\\"><div style=\\"text-align: center\\">hi
</div></body></html>"
`;

exports[`generators/MinimalHTML algorithm handles trivial inputs 1`] = `
"<!DOCTYPE html><html><body style=\\"color:#303030;font-weight:400;white-space:pre-wrap;font-family:sans-serif\\"><ul style=\\"margin:0px;list-style-position:inside;padding-left:1.5rem\\"><li>hi
</li></ul></body></html>"
`;
